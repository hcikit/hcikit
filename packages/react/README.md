# @hcikit/react

This package is a package that allows you to create and run HCI experiments. Visit hcikit.github.io for more information.
