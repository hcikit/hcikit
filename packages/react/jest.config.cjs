/** @type {import('@ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "jsdom",
  moduleNameMapper: {
    "@hcikit/workflow(.*)$": "<rootDir>/../workflow/src$1",
  },
  testMatch: ["./src/**/*.[jt]s?(x)", "**/?(*.)+(spec|test).[t]s?(x)"],
  modulePathIgnorePatterns: ["./dist/", "./node_modules/"],
  testPathIgnorePatterns: ["./dist/", "./node_modules/"],
  transformIgnorePatterns: ["./dist/", "/node_modules/"],
  setupFilesAfterEnv: ["<rootDir>/setupTests.ts"],
  collectCoverageFrom: [
    "src/**/*.{ts,tsx}",
    "!/node_modules/",
    "!src/setupTests.ts",
    "!src/**/*.stories.{js,ts,tsx,jsx}",
  ],
};
