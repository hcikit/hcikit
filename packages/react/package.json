{
  "name": "@hcikit/react",
  "version": "1.0.0-alpha.14",
  "description": "An experiment runner with a set of pre-built experiment components for running hci experiments",
  "author": "blainelewis1",
  "license": "MIT",
  "repository": "hcikit/hcikit",
  "module": "dist/index.js",
  "main": "dist/index.js",
  "type": "module",
  "keywords": [
    "react",
    "hci",
    "experiment"
  ],
  "types": "./dist/index.d.ts",
  "exports": {
    "import": "./dist/index.js",
    "default": "./dist/index.js"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist",
    "README.md"
  ],
  "engines": {
    "node": ">=8",
    "npm": ">=5"
  },
  "scripts": {
    "test": "NODE_OPTIONS=--experimental-vm-modules yarn jest --env=jsdom",
    "test:watch": "NODE_OPTIONS=--experimental-vm-modules yarn jest --watch --env=jsdom --colors --verbose",
    "start": "tsc --watch",
    "build": "tsc --build --verbose",
    "prepack": "yarn run build",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "peerDependencies": {
    "@mui/icons-material": "^5.5.0",
    "@mui/material": "^5.5.0",
    "@mui/styles": "^5.5.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "dependencies": {
    "@emotion/react": "^11.11.1",
    "@emotion/styled": "^11.11.0",
    "@hcikit/workflow": "workspace:*",
    "core-js": "^3.32.2",
    "lodash-es": "^4.17.21",
    "prop-types": "^15.8.1",
    "react-error-boundary": "^4.0.11",
    "react-is": "^18.2.0",
    "react-markdown": "^8.0.7",
    "serialize-error": "^11.0.2"
  },
  "gitHead": "8d0f2af41121f70996e8b5803d006832b147fd86",
  "devDependencies": {
    "@emotion/react": "^11.11.1",
    "@emotion/styled": "^11.11.0",
    "@mui/icons-material": "^5.14.11",
    "@mui/material": "^5.14.11",
    "@mui/styles": "^5.14.11",
    "@testing-library/jest-dom": "^6.1.3",
    "@testing-library/react": "^14.0.0",
    "@types/jest": "^29.5.5",
    "@types/react": "^18.2.23",
    "jest-diff": "^29.7.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-scripts": "^5.0.1",
    "typescript": "^5.2.2"
  }
}
