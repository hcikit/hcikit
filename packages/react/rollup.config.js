import babel from "rollup-plugin-babel";
import commonjs from "@rollup/plugin-commonjs";
import resolve from "@rollup/plugin-node-resolve";
import pkg from "./package.json";
import postcss from "rollup-plugin-postcss";

export default {
  input: "src/index.js",
  output: [
    {
      file: pkg.main,
      format: "cjs",
      sourcemap: true,
    },
    {
      file: pkg.module,
      format: "es",
      sourcemap: true,
    },
  ],
  external: Object.keys(pkg.peerDependencies || {}),
  plugins: [
    resolve(),
    commonjs({
      exclude: [
        "src/**",
        "../../node_modules/lodash-es/**",
        "../../node_modules/symbol-observable/es/**",
      ],
      include: ["../../node_modules/**"],
      namedExports: {
        "../../node_modules/react-redux/node_modules/react-is/index.js": [
          "isValidElementType",
          "isContextConsumer",
        ],
        "../../node_modules/@material-ui/utils/node_modules/react-is/index.js": [
          "ForwardRef",
          "Memo",
        ],
        "../../node_modules/react-dom/index.js": [
          "findDOMNode",
          "unstable_batchedUpdates",
        ],
        "../../node_modules/react-is/index.js": [
          "isValidElementType",
          "isElement",
          "ForwardRef",
          "typeOf",
          "Memo",
          "isContextConsumer",
        ],
        "../../node_modules/prop-types/index.js": ["element", "elementType"],
      },
    }),
    babel({
      exclude: "../../node_modules/**",
    }),
    postcss(),
  ],
};
