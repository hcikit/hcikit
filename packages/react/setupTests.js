import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

let error = console.error;

console.error = function (message) {
  error.apply(console, arguments); // keep default behaviour
  throw message instanceof Error ? message : new Error(message);
};
