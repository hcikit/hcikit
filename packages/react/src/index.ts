import React, { PropsWithChildren, ReactElement } from "react";
import { ControlFunctions } from "./core/Experiment";
import * as allTasks from "./tasks";

export * from "./tasks";
export * from "./components";

export { allTasks };

export { withGridItem } from "./GridLayout";

export { default as GridLayout } from "./GridLayout";
export { default } from "./core/Experiment";
export * from "./core/Experiment";

export interface FunctionTask<P = {}>
  extends React.FunctionComponent<P & ControlFunctions> {
  (props: PropsWithChildren<P & ControlFunctions>, context?: any): ReactElement<
    any,
    any
  > | null;
}

export abstract class Task<P = {}, S = {}, SS = any> extends React.Component<
  P & ControlFunctions,
  S,
  SS
> {}
