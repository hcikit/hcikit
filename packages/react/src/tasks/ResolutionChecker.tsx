import { Typography } from "@material-ui/core";
import { noop } from "lodash";
import React, { useEffect, useMemo, useState } from "react";
import ErrorOverlay from "../components/ErrorOverlay";
import { ControlFunctions } from "../core/Experiment";

// TODO: there is a bug where if you fix one of the resoltuion problems, the view doesn't update. It's because the state doesn';t change. Need to write a repro test.

const ResolutionChecker: React.FunctionComponent<{
  minXResolution?: number;
  maxXResolution?: number;
  minYResolution?: number;
  maxYResolution?: number;
  log?: ControlFunctions["log"];
}> = ({
  minXResolution,
  maxXResolution,
  minYResolution,
  maxYResolution,
  log = noop,
}) => {
  const [validResolution, setValidResolution] = useState(true);

  let conditions = useMemo(
    () => [
      {
        condition: () =>
          minXResolution ? window.innerWidth >= minXResolution : true,
        dimension: "minXResolution",
        message:
          "Your window is too small. The width must be at least " +
          minXResolution +
          "px.",
      },
      {
        condition: () =>
          maxXResolution ? window.innerWidth <= maxXResolution : true,
        dimension: "maxXResolution",
        message:
          "Your window is too big. The width must be smaller than " +
          maxXResolution +
          "px.",
      },
      {
        condition: () =>
          minYResolution ? window.innerHeight >= minYResolution : true,
        dimension: "minYResolution",
        message:
          "Your window is too small. The height must be at least " +
          minYResolution +
          "px.",
      },
      {
        condition: () =>
          maxYResolution ? window.innerHeight <= maxYResolution : true,
        dimension: "maxYResolution",
        message:
          "Your window is too big. The height must be smaller than " +
          maxYResolution +
          "px.",
      },
    ],
    [maxXResolution, maxYResolution, minXResolution, minYResolution]
  );

  useEffect(() => {
    function handleResize() {
      setValidResolution(conditions.every(({ condition }) => condition()));
      log({
        type: "resize",
        valid: conditions.every(({ condition }) => condition()),
        width: window.innerWidth,
        height: window.innerHeight,
        failingDimensions: conditions
          .filter(({ condition }) => !condition())
          .map(({ dimension }) => dimension),
      });
    }

    handleResize();

    window.addEventListener("resize", handleResize);
    return () => window.removeEventListener("resize", handleResize);
  }, [conditions, log]);

  if (!validResolution) {
    return (
      <ErrorOverlay>
        {conditions
          .filter((condition) => !condition.condition())
          .map(({ message }, i) => (
            <Typography key={i}>{message}</Typography>
          ))}
      </ErrorOverlay>
    );
  } else {
    return null;
  }
};

export default ResolutionChecker;
