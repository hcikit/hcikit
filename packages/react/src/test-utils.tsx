import { render, RenderResult } from "@testing-library/react";

import {
  ConfigContext,
  ConfigMutatorContext,
  ControlFunctions,
} from "./core/Experiment.js";
import { Configuration, ExperimentIndex } from "@hcikit/workflow";
import React from "react";
import { BasePersistence } from "./persistence/index.js";

class DoNothingPersistence extends BasePersistence {
  load(): Promise<Configuration | undefined> {
    return Promise.resolve(undefined);
  }
  save(state: Configuration): Promise<void> {
    return Promise.resolve();
  }
  clear(): Promise<void> {
    return Promise.resolve();
  }
  init(): Promise<void> {
    return Promise.resolve();
  }
}

const identities: ControlFunctions = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  advance: (_index?: ExperimentIndex) => {
    // do nothing.
  },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  log: (_log: unknown) => {
    // do nothing.
  },
  modify: (
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _modifiedConfig: Record<string, unknown>,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _index?: ExperimentIndex
  ) => {
    // do nothing.
  },
  persistence: new DoNothingPersistence(),
};

export const renderWithProvider = (
  Component: React.ReactElement,
  funcs: Partial<ControlFunctions> = identities,
  configuration: Configuration = {}
): RenderResult => {
  return render(
    <ConfigContext.Provider value={configuration}>
      <ConfigMutatorContext.Provider value={{ ...identities, ...funcs }}>
        {Component}
      </ConfigMutatorContext.Provider>
    </ConfigContext.Provider>
  );
};
