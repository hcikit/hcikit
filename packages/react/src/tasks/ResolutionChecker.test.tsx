import React from "react";
import ResolutionChecker from "./ResolutionChecker.js";
import { fireEvent, screen } from "@testing-library/react";

import { renderWithProvider } from "../test-utils.js";

let props = {
  maxXResolution: 1000,
  maxYResolution: 1000,
  minYResolution: 500,
  minXResolution: 500,
};

describe("ResolutionChecker", () => {
  beforeEach(() => {
    Object.assign(window, { innerWidth: 1000, innerHeight: 1000 });
  });

  it("renders nothing when correct", () => {
    let { container } = renderWithProvider(<ResolutionChecker {...props} />);
    expect(container).toBeEmptyDOMElement();
  });

  it("renders error for x too small", () => {
    Object.assign(window, { innerWidth: 499, innerHeight: 1000 });

    renderWithProvider(<ResolutionChecker {...props} />);

    screen.getByText(
      "Your window is too small. The width must be at least 500px."
    );
  });

  it("renders error for x too big", () => {
    Object.assign(window, { innerWidth: 1001, innerHeight: 1000 });
    renderWithProvider(<ResolutionChecker {...props} />);

    screen.getByText(
      "Your window is too big. The width must be smaller than 1000px."
    );
  });

  it("renders error for y too small", () => {
    Object.assign(window, { innerWidth: 1000, innerHeight: 499 });
    renderWithProvider(<ResolutionChecker {...props} />);

    screen.getByText(
      "Your window is too small. The height must be at least 500px."
    );
  });

  it("renders error for y too big", () => {
    Object.assign(window, { innerWidth: 1000, innerHeight: 1001 });
    renderWithProvider(<ResolutionChecker {...props} />);

    screen.getByText(
      "Your window is too big. The height must be smaller than 1000px."
    );
  });

  it("renders multiple errors", () => {
    Object.assign(window, { innerWidth: 1001, innerHeight: 1001 });
    const elem = renderWithProvider(<ResolutionChecker {...props} />);

    screen.getByText(
      "Your window is too big. The height must be smaller than 1000px."
    );
    screen.getByText(
      "Your window is too big. The width must be smaller than 1000px."
    );
    expect(elem.asFragment()).toMatchSnapshot();
  });

  it("detects problems after resize", () => {
    Object.assign(window, { innerWidth: 1000, innerHeight: 1000 });

    const { container } = renderWithProvider(<ResolutionChecker {...props} />);

    expect(container).toBeEmptyDOMElement();

    Object.assign(window, { innerWidth: 1000, innerHeight: 499 });

    fireEvent(window, new Event("resize"));

    screen.getByText(
      "Your window is too small. The height must be at least 500px."
    );
  });

  it("logs properly", () => {
    Object.assign(window, { innerWidth: 1001, innerHeight: 1001 });
    const log = jest.fn();

    renderWithProvider(<ResolutionChecker {...props} log={log} />);

    expect(log).toHaveBeenLastCalledWith({
      type: "resize",
      width: 1001,
      height: 1001,
      failingDimensions: ["maxXResolution", "maxYResolution"],
      valid: false,
    });

    Object.assign(window, { innerWidth: 499, innerHeight: 499 });
    fireEvent(window, new Event("resize"));

    expect(log).toHaveBeenLastCalledWith({
      type: "resize",
      width: 499,
      height: 499,
      failingDimensions: ["minXResolution", "minYResolution"],
      valid: false,
    });

    Object.assign(window, { innerWidth: 499, innerHeight: 501 });
    fireEvent(window, new Event("resize"));

    expect(log).toHaveBeenLastCalledWith({
      type: "resize",
      width: 499,
      height: 501,
      failingDimensions: ["minXResolution"],
      valid: false,
    });

    Object.assign(window, { innerWidth: 501, innerHeight: 501 });
    fireEvent(window, new Event("resize"));

    expect(log).toHaveBeenLastCalledWith({
      type: "resize",
      width: 501,
      height: 501,
      failingDimensions: [],
      valid: true,
    });
  });
});
