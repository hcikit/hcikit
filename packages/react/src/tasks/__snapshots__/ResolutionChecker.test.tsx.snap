// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolutionChecker renders multiple errors 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM kCbjDY"
  >
    <div
      class="MuiPaper-root sc-gtsrHT hponZY MuiPaper-elevation1 MuiPaper-rounded"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Your window is too big. The width must be smaller than 1000px.
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Your window is too big. The height must be smaller than 1000px.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
